provenance.jq

```diff
@@ -0,0 +1,80 @@
+# input: "build" object (with "buildId" top level key)
+# output: array of image tags
+def tags:
+  .source.arches[].tags[],
+  .source.arches[].archTags[],
+  .build.img
+;
+
+# input: "build" object (with "buildId" top level key)
+# output: purl platform query string
+def platform_string:
+  .source.arches[].platformString | gsub("/"; "%2F")
+;
+
+# input: "tags" object with image digest and platform arguments
+# output: json object for in-toto provenance subject field
+def subjects($platform; $digest):
+  {
+      "name": ("pkg:docker/" + . + "?platform=" + $platform),
+      "digest": {
+        "sha256": $digest
+      }
+  }
+;
+
+# input: GITHUB context argument
+# output: json object for in-toto provenance external parameters field
+def github_external_parameters($context):
+($context.workflow_ref | gsub( $context.repository + "/"; "")) as $workflowPathRef |
+{
+  inputs: $context.event.inputs,
+  workflow: {
+    ref: ($workflowPathRef | split("@")[1]),
+    repository: ($context.server_url + "/" + $context.repository),
+    path: ($workflowPathRef | split("@")[0]),
+    digest: {sha256: $context.workflow_sha}
+  }
+}
+;
+
+# input: GITHUB context argument
+# output: json object for in-toto provenance internal parameters field
+def github_internal_parameters($context):
+{
+  github: {
+    event_name: $context.event_name,
+    repository_id: $context.repository_id,
+    repository_owner_id: $context.repository_owner_id,
+  }
+}
+;
+
+# input: "tags" object with platform, image digest and GITHUB context arguments
+# output: json object for in-toto provenance statement
+def github_actions_provenance($platform; $digest; $context):
+{
+  _type: "https://in-toto.io/Statement/v1",
+  subject: . | map(subjects($platform; $digest)),
+  predicateType: "https://slsa.dev/provenance/v1",
+  predicate: {
+        buildDefinition: {
+            buildType: "https://slsa-framework.github.io/github-actions-buildtypes/workflow/v1",
+            externalParameters: github_external_parameters($context),
+            internalParameters: github_internal_parameters($context),
+            resolvedDependencies: [{
+                uri: ("git+"+$context.server_url+"/"+$context.repository+"@"+$context.ref),
+                digest: { "gitCommit": $context.sha }
+            }]
+        },
+        runDetails: {
+            builder: {
+                id: ($context.server_url+"/"+$context.workflow_ref),
```

Comment:
Yes, I have been going back and forth on what to do here. One option would be to tag the workflow commit and use the tag in the `workflow_dispatch` such that `$context.workflow_ref` is versioned via tagging. But using tags didn't seem like the best approach.

Another option would be to include the commit here. `($context.server_url+"/"+$context.workflow_ref+"@"+$context.workflow_sha)` but this will effectively make any change to `build.yml` ripple into the verification side as we need to be able to identify the builder and a new `sha` would be a new builder. Maybe that is acceptable, how often do we expect `build.yml` to change?

Reply:
At this point in time `($context.server_url+"/"+$context.workflow_ref)` will uniquely identify this workflow based on our build evaluation until we make a change to `build.yml` that invalidates that.

Reply:
After a discussion about this, I believe we decided to leave this as-is currently as the ID here will point to the analysis for the build platform as implemented after changes to meet SLSA build level 3. At any point there is a change the invalidates the security properties of the build platform then we would rev the builder ID accordingly (and rotate the public key bound to it).

Reply:
Yeah, that seems OK -- I don't love that we might forget / miss a change that _should've_ changed this, but it's fair.  Perhaps we can add a comment here to remind us when we review in the future that this field needs careful review and that _some_ changes mean we should both change it and rotate the key? (especially noting here that if it ever changes, we should also rotate the key :eyes:)

Reply:
added
```sh
# builder.id identifies the transitive closure of the trusted build platform evalution.
# any changes that alter security properties or build level must update this ID and rotate the signing key.
# https://slsa.dev/spec/v1.0/provenance#builder
```

we'll want to add a comment to the overall `build.yml` as well but will be in a separate PR